### ISSUE TITLE
BigQuery QP should replace `temporal-type` stuff with the typed HoneySQL stuff used in the main `:sql` QP

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#19759]
BigQuery is extremely fussy and has different versions of various functions for different temporal types, even for `datetime` and `timestamp` (see #11222). In  #11492 I fixed this by adding new `temporal-type` and `with-temporal-type` multimethods to the BigQuery QP namespace that can be used to determine the temporal type of something and add that information as well. 

Later on to fix #11502, I created a more generalized version of the stuff I did in #11492 that is used in the main SQL QP namespace itself -- #15115.

The bespoke BigQuery stuff can now be removed and rewritten to use the methods introduced in #15115.

**Comment by willbryant** [comment]
Will this get done as part of the hsql2 migration @camsaul?

Reason I ask is there's some remaining issues with field filters on expressions like day-of-week (eg. "Exclude > Monday"), but it doesn't seem worth trying to fix them in the current code if it's all about to get rewritten anyway.

**Comment by snoe** [comment]
Spent a couple days looking at this. It's a pretty huge undertaking as the `temporal-type` stuff is used both on honeysql elements and on mbql elements. It also relies heavily on metadata that is quite hard to follow and debug. Tests are also written depending on these internal details and so it hard to tell what is necessary for correct queries or just an implementation detail.

The stuff in the main `:sql` QP is better but there are some things missing, like a good place to inject timezone, coerce java types, or use custom coersion functions instead of `cast`.

My opinion is that rewriting the driver qp path from scratch would probably be a faster, safer, and cleaner path than try to steer the existing code. Maybe someone else takes this up in the future.
